cmd/yggdrasil/chuser_unix.go

```diff
@@ -0,0 +1,87 @@
+//go:build aix || darwin || dragonfly || freebsd || linux || netbsd || openbsd || solaris
+// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	osuser "os/user"
+	"strconv"
+	"strings"
+	"syscall"
+)
+
+func chuser(user string) error {
+	group := ""
+	if i := strings.IndexByte(user, ':'); i >= 0 {
+		user, group = user[:i], user[i+1:]
+	}
+
+	u := (*osuser.User)(nil)
+	g := (*osuser.Group)(nil)
+
+	if user != "" {
+		if _, err := strconv.ParseUint(user, 10, 32); err == nil {
+			u, err = osuser.LookupId(user)
+			if err != nil {
+				return fmt.Errorf("failed to lookup user by id %q: %v", user, err)
+			}
+		} else {
+			u, err = osuser.Lookup(user)
+			if err != nil {
+				return fmt.Errorf("failed to lookup user by name %q: %v", user, err)
+			}
+		}
+	}
+	if group != "" {
+		if _, err := strconv.ParseUint(group, 10, 32); err == nil {
+			g, err = osuser.LookupGroupId(group)
+			if err != nil {
+				return fmt.Errorf("failed to lookup group by id %q: %v", user, err)
+			}
+		} else {
+			g, err = osuser.LookupGroup(group)
+			if err != nil {
+				return fmt.Errorf("failed to lookup group by name %q: %v", user, err)
+			}
+		}
+	}
+
+	if g != nil {
+		gid, _ := strconv.ParseUint(g.Gid, 10, 32)
+		var err error
+		if gid < math.MaxInt {
```

Comment:
os/user returns ID strings that were "validated" with strconv.Atoi() which returns `int`, so this should always be true.

Reply:
had to dig deep to confirm this but indeed.
also could probably be `<=`.
feel free to make a PR if you care for this as this looks like something code being saner internally not something that'll lead to bugs.

Reply:
> feel free to make a PR if you care for this as this looks like something code being saner internally not something that'll lead to bugs.

Either that or a clean(er) `chuser_bsd.go` implementation, should I ever start using this option.

Point being:  Complexity hides bugs, so code ought to be as simple as possible.

To my eye, multiple Lookup*() and Parse*() are not as trivial to reason about as it could be.

Reply:
without digging deep into the source of `osuser.LookupId` one can't even know whether these ints were checked so I'm not sure complexity argument is even applicable here, it's more complex to verify whether the code is correct if there is no check.
my gut instinct when API doesn't tell stuff like returned ranges of ints to just check myself, because checking more usually catches more bugs than not checking stuff like returns.
IMO if check is eliminated, then strconv.Atoi() check behavior should be documented in comment, as otherwise it won't end up more simple.

Reply:
also uint32 vs int is potentially hell on its own due to int being 64bit on 64bit systems and strconv.Atoi() ending up verifying different on 32bit vs 64bit systems but honestly systems just don't use ints as large as these for UIDs/GIDs anyways so i can't really care abt this.

Reply:
https://github.com/yggdrasil-network/yggdrasil-go/pull/927/commits/87fa0a8936d554d3f885ee6f7f6d6647c8c75f65 apparently was someone's else's change too but I can't argue one way or another abt it.

Reply:
to satisfy faulty Code scanning / CodeQL warnings out of all the things

Reply:
> to satisfy faulty Code scanning / CodeQL warnings out of all the things

This looks related over at #1202:
![Screenshot 2024-11-04 at 04-05-21 Set groups when dropping privileges to not leak supplementary group access by klemensn · Pull Request #1202 · yggdrasil-network_yggdrasil-go](https://github.com/user-attachments/assets/5bcfa044-b7fc-49de-a354-c3e9e0bc98ce)

Reply:
> feel free to make a PR if you care for this as this looks like something code being saner internally not something that'll lead to bugs.

- `-user ''` and `-user ':'` are silently ignored and the process continues without error or change in privilege.
- `-user :group` change only the group, but not the user, contrary to what the usage says:
```
  -user string
    	user (and, optionally, group) to set UID/GID to
```

I did fix these now in #1203.